src/calcium/ca.jl

```diff
@@ -0,0 +1,1447 @@
+        Ptr{UInt8}, (Ref{ca}, Ref{CalciumField}), x, x.parent)
+   res = unsafe_string(cstr)
+   ccall((:flint_free, libflint), Nothing, (Ptr{UInt8},), cstr)
+
+   return res
+end
+
+function show(io::IO, x::ca)
+   print(io, native_string(x))
+end
+
+needs_parentheses(x::ca) = true
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+zero(C::CalciumField) = C()
+
+function one(C::CalciumField)
+   z = ca(C)
+   ccall((:ca_one, libcalcium), Nothing, (Ref{ca}, Ref{CalciumField}), z, C)
+   return z
+end
+
+###############################################################################
+#
+#   Random generation
+#
+###############################################################################
+
+function rand(C::CalciumField; depth::Int, bits::Int,
+                                            randtype::Symbol=:null)
+   state = _flint_rand_states[Threads.threadid()]
+   x = C()
+
+   depth = max(depth, 0)
+   bits = max(bits, 1)
+
+   if randtype == :null
+      ccall((:ca_randtest, libcalcium), Nothing,
+          (Ref{ca}, Ptr{Cvoid}, Int, Int, Ref{CalciumField}),
+                x, state.ptr, depth, bits, C)
+   elseif randtype == :rational
+      ccall((:ca_randtest_rational, libcalcium), Nothing,
+          (Ref{ca}, Ptr{Cvoid}, Int, Ref{CalciumField}),
+                x, state.ptr, bits, C)
+   elseif randtype == :special
+      ccall((:ca_randtest_special, libcalcium), Nothing,
+          (Ref{ca}, Ptr{Cvoid}, Int, Int, Ref{CalciumField}),
+                x, state.ptr, depth, bits, C)
+   else
+      error("randtype not defined")
+   end
+
+   check_special(x)
+   return x
+end
+
+###############################################################################
+#
+#   Arithmetic
+#
+###############################################################################
+
+function _isspecial(a::ca)
+   return (a.field & 3) != 0
+end
+
+# todo: distiguish unknown
+function check_special(a::ca)
+   if !a.parent.extended && _isspecial(a)
+      throw(DomainError(a, "Non-number result"))
+   end
+end
+
+function same_parent(a::ca, b::ca)
+   if a.parent == b.parent
+      return (a, b)
+   else
+      C = a.parent
+      r = C()
+      ccall((:ca_transfer, libcalcium), Nothing,
+         (Ref{ca}, Ref{CalciumField}, Ref{ca}, Ref{CalciumField}),
+         r, a.parent, b, b.parent)
+      check_special(r)
+      return (a, r)
+   end
+end
+
+function +(a::ca, b::ca)
+   a, b = same_parent(a, b)
+   C = a.parent
+   r = C()
+   ccall((:ca_add, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Ref{ca}, Ref{CalciumField}), r, a, b, C)
+   check_special(r)
+   return r
+end
+
+function +(a::ca, b::Int)
+   C = a.parent
+   r = C()
+   ccall((:ca_add_si, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Int, Ref{CalciumField}), r, a, b, C)
+   check_special(r)
+   return r
+end
+
+function +(a::ca, b::fmpz)
```

Comment:
Do we have generic code that switches the order of arguments if one ad hoc operation is available?

Reply:
No idea what's happening behind the scenes, but the reverse operation works and is tested. I will add the reverse functions anyway to make sure it's fast.